node/rate_limiter.js

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+
+var DEFAULT_SERVICE_RPS_LIMIT = 100;
+var DEFAULT_TOTAL_RPS_LIMIT = 1000;
+var DEFAULT_BUCKET_NUMBER = 20;
+
+function RateLimiterCounter(options) {
+    if (!(this instanceof RateLimiterCounter)) {
+        return new RateLimiterCounter(options);
+    }
+
+    var self = this;
+    self.index = 0;
+    self.rps = 0;
+    self.numOfBuckets = options.numOfBuckets;
+    self.buckets = [];
+    self.buckets.length = self.numOfBuckets;
+
+    // self.buckets is read/written in refresh,
+    // where read is always after write on a bucket.
+    self.buckets[0] = 0;
+    self.rpsLimit = options.rpsLimit;
+}
+
+RateLimiterCounter.prototype.refresh =
+function refresh() {
+    var self = this;
+
+    // update the sliding window
+    var next = (self.index + 1) % self.numOfBuckets;
+    if (self.buckets[next]) {
+        // offset the bucket being moved out
+        self.rps -= self.buckets[next];
+    }
+
+    assert(self.rps >= 0, 'rps should always be larger equal to 0');
+    self.index = next;
+    self.buckets[self.index] = 0;
```

Comment:
Note that this will still work, even for a densely packed array, when self.index === self.buckets.length, without turning it into a hashtable, making it unnecessary to declare the array size up front, or “pre-allocating” a memory region by assigning to the array length.